src/openloco/window.h

```diff
@@ -24,18 +43,82 @@ namespace openloco::ui
         string_id tooltip; // 0x0E
     };
 
+    struct scroll
+    {
+        uint16_t flags;
+        uint16_t h_left;
+        uint16_t h_right;
+        uint16_t h_thumb_left;
+        uint16_t h_thumb_right;
+        uint16_t v_top;
+        uint16_t v_bottom;
+        uint16_t v_thumb_top;
+        uint16_t v_thumb_bottom;
+    };
+
+    struct viewport
+    {
+        int16_t width;       // 0x00
+        int16_t height;      // 0x02
+        int16_t x;           // 0x04
+        int16_t y;           // 0x06
+        int16_t view_x;      // 0x08
+        int16_t view_y;      // 0x0A
+        int16_t view_width;  // 0x0C
+        int16_t view_height; // 0x0E
+        uint8_t zoom;        // 0x10
+        uint8_t pad_11;
+        uint16_t var_12; // 0x12, maybe flags
+    };
+
+    struct window;
+
+    struct window_event_list_t
+    {
+        uint32_t event_00;
+        uint32_t event_01;
+        uint32_t event_02;
+        uint32_t event_03;
+        void (*on_mouse_down)(window*, int8_t widget_index); // 0x10
+        uint32_t event_05;
+        uint32_t event_06;
+        uint32_t event_07;
+        uint32_t event_08;
+        uint32_t event_09;
+        uint32_t event_10;
+        uint32_t event_11;
+        uint32_t event_12;
+        uint32_t event_13;
+        uint32_t event_14;
+        uint32_t event_15;
+        uint32_t event_16;
+        void (*scroll_mouse_down)(window*, int16_t x, int16_t y, int8_t scroll_index); // 0x10
+        uint32_t event_18;
+        uint32_t event_19;
+        uint32_t event_20;
+        uint32_t event_21;
+        uint32_t event_22;
+        uint32_t event_23;
+        uint32_t event_24;
+        uint32_t event_25;
+        void (*invalidate)(window*);                   //0x68
+        void (*paint)(window*, gfx::drawpixelinfo_t&); // 0x6C
+        uint32_t event_28;
+    };
+
     struct window
     {
         union
         {
             uint8_t pad_all[0x88E];
             struct
             {
-                void* event_handlers;     // 0x00
-                void* viewport;           // 0x04
-                uint8_t pad_08[0x04];     // 0x08
-                uint32_t enabled_widgets; // 0x0C
-                uint8_t pad_10[0x2C - 0x10];
+                window_event_list_t* event_handlers; // 0x00
+                ui::viewport* viewport;              // 0x04
+                uint8_t pad_08[0x04];                // 0x08
+                uint64_t enabled_widgets;            // 0x0C
```

Comment:
I'm surprised to see these as `uint64_t`, considering the original game is 32-bits. OpenRCT2 also has them as a 32-bits value. Could you confirm they are indeed 64-bits?

Reply:
From OpenRCT2:
```
struct rct_window {
    rct_window_event_list* event_handlers;      // 0x000
    rct_viewport* viewport;     // 0x004
    uint64 enabled_widgets;     // 0x008
    uint64 disabled_widgets;    // 0x010
    uint64 pressed_widgets;     // 0x018
    uint64 hold_down_widgets;   // 0x020
    rct_widget* widgets;        // 0x028
```

Reply:
My mistake. I think I misremembered because some widget lists are defined with uint32, an inconsistency we should probably address:
```
src/openrct2-ui/windows/Guest.cpp:static constexpr const uint32 window_guest_page_enabled_widgets[]
src/openrct2-ui/windows/Park.cpp:static uint32 window_park_page_enabled_widgets[]
src/openrct2-ui/windows/Player.cpp:static uint32 window_player_page_enabled_widgets[]
src/openrct2-ui/windows/Research.cpp:static uint32 window_research_page_enabled_widgets[]
src/openrct2-ui/windows/Staff.cpp:static constexpr const uint32 window_staff_page_enabled_widgets[]
```